From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Sat, 17 Dec 2022 13:04:14 +0100
Subject: [PATCH] Avoid respawn desync of player metadata

== AT ==
public net.minecraft.world.entity.player.Player DATA_PLAYER_ABSORPTION_ID
public net.minecraft.world.entity.player.Player DATA_SCORE_ID

diff --git a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
index 37e193f57938047c8b886ed7d2816411392f94b4..f7d1425d102959f7a2f1da5585e2d3d2032a5abb 100644
--- a/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/src/main/java/net/minecraft/network/syncher/SynchedEntityData.java
@@ -152,6 +152,15 @@ public class SynchedEntityData {
         this.isDirty = true;
     }
     // CraftBukkit end
+    // Paper start
+    public <T> void markDirtyIfNotDefault(EntityDataAccessor<T> datawatcherobject) {
+        DataItem<T> item = this.getItem(datawatcherobject);
+        if (!item.isSetToDefault()) {
+            item.setDirty(true);
+            this.isDirty = true;
+        }
+    }
+    // Paper end
 
     public boolean isDirty() {
         return this.isDirty;
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index c0b0a7fdb75266a7064d54bda6441953184ecc64..14960c8e40f34047664ef7f2ddcb214bc65034f9 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1776,11 +1776,14 @@ public class ServerPlayer extends Player {
         if (alive) {
             this.getInventory().replaceWith(oldPlayer.getInventory());
             this.setHealth(oldPlayer.getHealth());
+            this.getEntityData().markDirtyIfNotDefault(LivingEntity.DATA_HEALTH_ID); // Paper
+            this.getEntityData().markDirtyIfNotDefault(Player.DATA_PLAYER_ABSORPTION_ID); // Paper
             this.foodData = oldPlayer.foodData;
             this.experienceLevel = oldPlayer.experienceLevel;
             this.totalExperience = oldPlayer.totalExperience;
             this.experienceProgress = oldPlayer.experienceProgress;
             this.setScore(oldPlayer.getScore());
+            this.getEntityData().markDirtyIfNotDefault(Player.DATA_SCORE_ID); // Paper
             this.portalEntrancePos = oldPlayer.portalEntrancePos;
         } else if (this.level.getGameRules().getBoolean(GameRules.RULE_KEEPINVENTORY) || oldPlayer.isSpectator()) {
             this.getInventory().replaceWith(oldPlayer.getInventory());
@@ -1788,11 +1791,12 @@ public class ServerPlayer extends Player {
             this.totalExperience = oldPlayer.totalExperience;
             this.experienceProgress = oldPlayer.experienceProgress;
             this.setScore(oldPlayer.getScore());
+            this.getEntityData().markDirtyIfNotDefault(Player.DATA_SCORE_ID); // Paper
         }
 
         this.enchantmentSeed = oldPlayer.enchantmentSeed;
         this.enderChestInventory = oldPlayer.enderChestInventory;
-        this.getEntityData().set(ServerPlayer.DATA_PLAYER_MODE_CUSTOMISATION, (Byte) oldPlayer.getEntityData().get(ServerPlayer.DATA_PLAYER_MODE_CUSTOMISATION));
+        this.getEntityData().markDirtyIfNotDefault(Player.DATA_PLAYER_MODE_CUSTOMISATION); // Paper
         this.lastSentExp = -1;
         this.lastSentHealth = -1.0F;
         this.lastSentFood = -1;
@@ -1800,7 +1804,9 @@ public class ServerPlayer extends Player {
         this.seenCredits = oldPlayer.seenCredits;
         this.enteredNetherPosition = oldPlayer.enteredNetherPosition;
         this.setShoulderEntityLeft(oldPlayer.getShoulderEntityLeft());
+        this.getEntityData().markDirtyIfNotDefault(Player.DATA_SHOULDER_LEFT); // Paper
         this.setShoulderEntityRight(oldPlayer.getShoulderEntityRight());
+        this.getEntityData().markDirtyIfNotDefault(Player.DATA_SHOULDER_RIGHT); // Paper
         this.setLastDeathLocation(oldPlayer.getLastDeathLocation());
     }
 
